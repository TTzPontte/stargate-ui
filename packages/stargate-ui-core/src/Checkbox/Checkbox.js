import PropTypes from "prop-types";
import React from 'react';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';

import Factory from '../Factory';

const styles = (theme) => {
  const {
    palette,
    radius,
    spacing,
    border,
    active,
    mode,
  } = theme;

  const getColor = (color) => (
    palette?.[color][mode].color
  );

  const checkbox = {
    width: '100%',
    border: 0,
    borderRadius: '50%',
    backgroundColor: (props) => {
      const {
        disabled,
        readonly,
        color,
      } = props;

      return (disabled || readonly) && setLightness(.95, getColor(color));
    },
  };


  return {checkbox};
};

const Checkbox = (props) => {
  const {
    disabled,
    readonly,
    label,
    color = 'default',
    onChange = () => {},
    ...factoryProps
  } = props;

  const [
    {
      checkbox: classCheckbox,
      ...classes
    }
  ] = useStyles(styles, {
    disabled,
    readonly,
    color,
  });
  const className = clsx(Object.values(classes));
  console.log(classCheckbox)

  return (
      <Factory
        element="input"
        type="checkbox"
        className={classCheckbox}
        disabled={disabled}
        readOnly={readonly}
        {...factoryProps}
      />
  );
};

Checkbox.propTypes = {
  color: PropTypes.string,
  disabled: PropTypes.any,
  label: PropTypes.any,
  onChange: PropTypes.func,
  readonly: PropTypes.any,
  type: PropTypes.string
}

export default Checkbox;
