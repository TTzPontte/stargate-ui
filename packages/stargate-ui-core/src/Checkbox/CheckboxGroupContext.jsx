import React from 'react';

const CheckboxGroupContext = React.createContext({});

CheckboxGroupContext.displayName = 'CheckboxGroupContext';

export default CheckboxGroupContext;
