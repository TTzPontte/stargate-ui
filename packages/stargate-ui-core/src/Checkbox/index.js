export { default } from './CheckboxGroup';
export { default } from './Checkbox';
