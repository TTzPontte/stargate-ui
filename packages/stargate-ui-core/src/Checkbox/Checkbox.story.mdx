import {
  Meta,
  Canvas,
  Status,
  Story,
  withAction,
  Props,
} from '@pontte/stargate-ui-playground';

import Typography from '../Typography';
import { Instagram as SvgIconInstagram } from '../icons';
import Label from '../Label';

import Checkbox from './Checkbox';

<Meta title="Core/Form Elements/Checkbox" component={Checkbox} />

# Checkbox

<Status type="rc" />

## Basic Usage

<Canvas>
  <Story name="Basic Usage" parameters={{
    jest: ['Checkbox.spec.jsx'],
  }}>
    {
      () => (
        <>
          <Checkbox
            checkboxLabel="Checkbox"
            checked
          />
          <Checkbox
            checkboxLabel="Checkbox Disabled"
            checked
            disabled
          />
        </>
      )
    }
  </Story>
</Canvas>

<Props component={Checkbox} />

## Checkbox with Label

<Canvas>
  <Checkbox
    label="This is a default Label"
    checkboxLabel="Checkbox"
    checked
  />
  <Checkbox
    label={
      <Label>
        <SvgIconInstagram />
        This is a Customized Label with Icon
      </Label>
    }
    checkboxLabel="Checkbox"
    checked
    disabled
  />
</Canvas>

## Colors State

<Canvas>
  <Checkbox
    color="success"
    checked={false}
  />
  <Checkbox
    color="warning"
    checked
  />
  <Checkbox
    color="error"
    checked={false}
  />
  <Checkbox
    color="info"
    checked
  />
</Canvas>
