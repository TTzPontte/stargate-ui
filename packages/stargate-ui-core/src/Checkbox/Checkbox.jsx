import React from 'react';
import PropTypes from "prop-types";
import clsx from 'clsx';

import { useStyles } from '@pontte/stargate-ui-styles';
import Label from '../Label';
import Typography from '../Typography';
import Factory from '../Factory';
import CheckboxGroupContext from './CheckboxGroupContext';

const styles = (theme) => {
  const {
    palette,
    radius,
    mode,
  } = theme;

  const { setLightness } = palette;

  const checkbox = {
    position: 'absolute',
    width: 0,
    height: 0,
    userSelect: 'none',
    opacity: 0,
  };

  const checkboxDecoration = {
    cursor: 'pointer',
    content: '""',
    display: 'inline-block',
    position: 'relative',
    minWidth: 20,
    minHeight: 20,
    borderRadius: radius(),
    backgroundColor: palette.lighter,
    margin: 0,
    overflow: 'hidden',
    border: [
      1,
      'solid',
      palette.default[mode].color,
    ],
    '$checkbox:checked': {
      background: palette.default[mode].color,
    },
    '$checkbox:checked:disabled ~ &:after, $checkbox:checked[readonly] ~ &:after': {
      borderBottomColor: palette.default[mode].color,
      borderRightColor: palette.default[mode].color,
    },
    '$checkbox:disabled ~ &, $checkbox[readonly] ~ &': {
      background: setLightness(.95, palette.default[mode].color),
    },
    '$checkbox:checked:disabled ~ &, $checkbox:checked[readonly] ~ &': {
      borderColor: palette.default[mode].color,
      background: setLightness(.95, palette.default[mode].color),
    },
    '&:after': {
      content: '""',
      display: 'inline-block',
      '$checkbox:checked ~ &': {
        top: '45%',
        left: '50%',
        transform: 'translate(-50%, -50%) rotate(45deg)',
        width: '32%',
        height: '56%',
        position: 'absolute',
        borderBottom: [2, 'solid', palette.lighter],
        borderRight: [2, 'solid', palette.lighter],
        borderTopLeftRadius: '2px',
        borderTopRightRadius: '2px',
        borderBottomLeftRadius: '2px',
        borderBottomRightRadius: '1px',
      },
    },
  };

  const checkboxDecorationSuccess = {
    '$checkbox:checked ~ &': {
      background: palette.success[mode].color,
      borderColor: palette.success[mode].color,
    },
  };

  const checkboxDecorationWarning = {
    '$checkbox:checked ~ &': {
      background: palette.warning[mode].color,
      borderColor: palette.warning[mode].color,
    },
  };

  const checkboxDecorationError = {
    '$checkbox:checked ~ &': {
      background: palette.error[mode].color,
      borderColor: palette.error[mode].color,
    },
  };

  return {
    checkbox,
    checkboxDecoration,
    checkboxDecorationSuccess,
    checkboxDecorationWarning,
    checkboxDecorationError,
  };
};

const Checkbox = React.forwardRef((props, ref) => {
  const {
    label,
    value,
    color: inheritedColor,
    checked: inheritedChecked,
    disabled: inheritedDisabled,
    readOnly: inheritedReadOnly,
    className: inheritedClassName,
    ...factoryProps
  } = props;

  const inputInner = React.useRef(ref);

  const {
    name,
    onChange,
    value: checkboxGroupValue,
    color: checkboxGroupColor,
    disabled: checkboxGroupDisabled,
    readOnly: checkboxGroupReadonly,
  } = React.useContext(CheckboxGroupContext);

  /**
   * Inherited properties has priority over @func CheckboxGroupContext
   */
  const color = inheritedColor || checkboxGroupColor || 'success';
  const disabled = inheritedDisabled ?? checkboxGroupDisabled;
  const readOnly = inheritedReadOnly ?? checkboxGroupReadonly;
  const checked = value && checkboxGroupValue?.length ? checkboxGroupValue.includes(value) : inheritedChecked;

  const handleChange = (event) => {
    if (disabled || readOnly) {
      event.preventDefault();
      return;
    }

    if (onChange) {
      onChange(event, [value]);
    }
  };

  const [{ checkbox: classCheckbox, ...classes }] = useStyles(styles);
  const classCheckboxDecoration = clsx(
    classes.checkboxDecoration,
    {
      [classes.checkboxDecorationSuccess]: color === 'success',
      [classes.checkboxDecorationError]: color === 'error',
      [classes.checkboxDecorationWarning]: color === 'warning',
    },
    inheritedClassName,
  );

  return (
    <Factory display="flex">
      <Label>
        <Factory
          {...factoryProps}
          disabled={disabled}
          readOnly={readOnly}
          checked={checked}
          ref={inputInner}
          element="input"
          type="checkbox"
          className={classCheckbox}
          onChange={handleChange}
          value={value}
          name={name}
        />

        <Factory
          element="span"
          className={classCheckboxDecoration}
          marginX={1}
        />

        <Typography
          variant="body"
          element="span"
          gutter={0}
        >
          {label}
        </Typography>
      </Label>
    </Factory>
  );
});

Checkbox.displayName = 'Checkbox';

Checkbox.propTypes = {
  /**
   * Add input value.
   *
   * **@default** `undefined`
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Disables button and add disabled CSS style.
   *
   * **@default** `undefined`
   */
  disabled: PropTypes.bool,
  /**
   * Add readonly CSS style.
   *
   * **@default** `undefined`
   */
  readonly: PropTypes.bool,
  /**
   * Add a label description for each radio.
   *
   * **@default** `undefined`
   */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]).isRequired,
  /**
   * Add color style.
   *
   * **@default** `success`
   */
  color: PropTypes.oneOf([
    'success',
    'warning',
    'error'
  ]),
};

export default Checkbox;
