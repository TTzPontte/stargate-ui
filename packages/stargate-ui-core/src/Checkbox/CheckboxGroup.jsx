import React from 'react';
import PropTypes from 'prop-types';

import CheckboxGroupContext from './CheckboxGroupContext';

const CheckboxGroup = (props) => {
  const {
    children,
    name,
    checkboxGroupValues,
    disabled,
    readOnly,
    color,
    onChange = () => {},
  } = props;

  const handleChange = React.useCallback(onChange, [onChange]);
  const context = React.useMemo(() => ({
    name,
    checkboxGroupValues,
    disabled,
    readOnly,
    color,
    onChange: handleChange,
  }), [
    name,
    checkboxGroupValues,
    disabled,
    readOnly,
    color,
    handleChange,
  ]);
  return (
    <CheckboxGroupContext.Provider
      value={context}
      children={children}
    />
  );
};

CheckboxGroup.displayName = 'CheckboxGroup';

CheckboxGroup.propTypes = {
  /**
   * Disables button and add disabled CSS style to Checkbox's components.
   *
   * **@default** `undefined`
   */
  disabled: PropTypes.bool,
  /**
   * Add readonly CSS style to Checkbox's components.
   *
   * **@default** `undefined`
   */
  readOnly: PropTypes.bool,
  /**
   * Add component dependencies.
   */
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.instanceOf(CheckboxGroup),
  ]).isRequired,
  /**
   * Add color style to Checkbox's components.
   *
   * **@default** `success`
   */
  color: PropTypes.oneOf([
    'success',
    'warning',
    'error'
  ]),
  /**
   * Specify a name for the CheckboxGroup.
   *
   * **@default** `undefined`
   */
  name: PropTypes.string.isRequired,
  // /**
  //  * Values of the chosen Checkbox.
  //  *
  //  * @default `''`
  //  */
  // value: PropTypes.string,
    /**
   * Values of the chosen Checkbox.
   *
   * @default `[]`
   */
  checkboxGroupValues: PropTypes.array,
  /**
   * Trigger when `Checkbox` is changed.
   *
   * @param {object} event Default on change event
   * @param {array} value Values of the chosen Checkbox
   *
   * @default `() => {}`
   */
  onChange: PropTypes.func,
};

export default CheckboxGroup;
