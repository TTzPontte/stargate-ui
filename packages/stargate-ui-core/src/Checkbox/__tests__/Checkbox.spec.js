import React from 'react';

import Checkbox from '../Checkbox';

describe('Checkbox', () => {
  it('should render properly', () => {
    const { container } = render(
      <Checkbox label="Label" />,
    );

    expect(container).toMatchSnapshot();
  });
});
