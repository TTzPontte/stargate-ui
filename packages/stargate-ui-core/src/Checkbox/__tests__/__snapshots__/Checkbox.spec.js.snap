// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render properly 1`] = `
<div>
  <div
    class="stargate-ui-factory-0-2-6 stargate-ui-factory-d0-0-2-8"
  >
    <label
      class="stargate-ui-factory-0-2-18 stargate-ui-factory-d0-0-2-20 stargate-ui-typography-0-2-11 stargate-ui-typography-d0-0-2-15 stargate-ui-label-0-2-10 stargate-ui-typographyFluid-0-2-14 stargate-ui-typographyFluid-d2-0-2-17"
    >
      <input
        checkboxlabel="Checkbox"
        class="stargate-ui-factory-0-2-22 stargate-ui-factory-d0-0-2-24 stargate-ui-checkbox-0-2-1"
        type="checkbox"
        value=""
      />
      <span
        class="stargate-ui-factory-0-2-26 stargate-ui-factory-d0-0-2-28 stargate-ui-checkboxDecoration-0-2-2 stargate-ui-checkboxDecorationSuccess-0-2-3"
      />
      <span
        class="stargate-ui-factory-0-2-37 stargate-ui-factory-d0-0-2-39 stargate-ui-typography-0-2-30 stargate-ui-typography-d0-0-2-34 stargate-ui-typographyFluid-0-2-33 stargate-ui-typographyFluid-d2-0-2-36"
      >
        Label
      </span>
    </label>
  </div>
</div>
`;
